src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/CompletionSource.cs

```diff
@@ -74,10 +74,13 @@ internal CompletionSource(ITextView textView, IThreadingContext threadingContext
                 _textView.Options.GlobalOptions.SetOptionValue(NonBlockingCompletionEditorOption, true);
             }
 
-            if (!triggerLocation.Snapshot.TextBuffer.Properties.ContainsProperty(PotentialCommitCharacters))
-            {
-                triggerLocation.Snapshot.TextBuffer.Properties.AddProperty(PotentialCommitCharacters, service.GetRules().DefaultCommitCharacters);
-            }
+            var commitCharacters = service.GetRules().DefaultCommitCharacters;
+
+            // In case of calls with multiple completion services for the same view (e.g. TypeScript and C#), we should merge potential commit characters together.
+            _textView.Properties[PotentialCommitCharacters] =
+                _textView.Properties.TryGetProperty(PotentialCommitCharacters, out ImmutableHashSet<char> savedCommitCharacters)
+                ? savedCommitCharacters.Union(commitCharacters)
```

Comment:
in 99% of cases we will be doing this union. It's on typing hot path, and additionally it uses LINQ. This won't be good for perf. Can we optimize this?